src/libAtomVM/externalterm.c

```diff
@@ -54,22 +57,217 @@ term externalterm_to_term(const void *external_term, Context *ctx)
+    if (UNLIKELY(memory_ensure_free(ctx, size_in_terms + 1) != MEMORY_GC_OK)) {
+        fprintf(stderr, "Unable to ensure %i free words in heap\n", size_in_terms);
+        return term_invalid_term();
+    }
+    //
+    // create and return the binary
+    //
+    term binary = term_from_literal_binary((void *)buf, len, ctx);
+    free(buf);
+    return binary;
+}
+
+static size_t compute_external_size(Context *ctx, term t)
+{
+    return serialize_term(ctx, NULL, t);
+}
+
+static int serialize_term(Context *ctx, uint8_t *buf, term t)
+{
+    if (term_is_uint8(t)) {
+        if (!IS_NULL_PTR(buf)) {
+            buf[0] = SMALL_INTEGER_EXT;
+            buf[1] = term_to_uint8(t);
+        }
+        return 2;
+
+    } else if (term_is_integer(t)) {
+        if (!IS_NULL_PTR(buf)) {
+            int32_t val = term_to_int32(t);
+            buf[0] = INTEGER_EXT;
+            WRITE_32_UNALIGNED(buf + 1, val);
+        }
+        return 5;
+
+    } else if (term_is_atom(t)) {
+        AtomString atom_string = globalcontext_atomstring_from_term(ctx->global, t);
+        size_t atom_len = atom_string_len(atom_string);
+        if (!IS_NULL_PTR(buf)) {
+            buf[0] = ATOM_EXT;
+            WRITE_16_UNALIGNED(buf + 1, atom_len);
+            int8_t *atom_data = (int8_t *) atom_string_data(atom_string);
+            for (size_t i = 3;  i < atom_len + 3;  ++i) {
+                buf[i] = (int8_t) atom_data[i - 3];
+            }
+        }
+        return 3 + atom_len;
+
+    } else if (term_is_tuple(t)) {
+        size_t arity = term_get_tuple_arity(t);
+        if (arity > 255) {
+            fprintf(stderr, "Tuple arity greater than 255: %zu\n", arity);
+            abort();
+        }
+        if (!IS_NULL_PTR(buf)) {
+            buf[0] = SMALL_TUPLE_EXT;
+            buf[1] = (int8_t) arity;
+        }
+        size_t k = 2;
+        for (size_t i = 0;  i < arity;  ++i) {
+            term e = term_get_tuple_element(t, i);
+            k += serialize_term(ctx, IS_NULL_PTR(buf) ? NULL : buf + k, e);
+        }
+        return k;
+
+    } else if (term_is_nil(t)) {
+        if (!IS_NULL_PTR(buf)) {
+            buf[0] = NIL_EXT;
+        }
+        return 1;
+
+    } else if (term_is_string(t)) {
+        if (!IS_NULL_PTR(buf)) {
+            buf[0] = STRING_EXT;
+        }
+        size_t len = 0;
+        size_t k = 3;
+        term i = t;
+        while (!term_is_nil(i)) {
+            term e = term_get_list_head(i);
+            if (!IS_NULL_PTR(buf)) {
+                *(buf + k) = term_to_uint8(e);
+            }
+            ++k;
+            i = term_get_list_tail(i);
+            ++len;
+        }
+        if (!IS_NULL_PTR(buf)) {
+            WRITE_16_UNALIGNED(buf + 1, len);
+        }
+        return k;
+
+    } else if (term_is_list(t)) {
+        if (!IS_NULL_PTR(buf)) {
+            buf[0] = LIST_EXT;
+        }
+        size_t len = 0;
+        size_t k = 5;
+        term i = t;
+        while (!term_is_nil(i)) {
```

Comment:
We need to support also improper lists: `while (term_is_nonempty_list(i))`

Reply:
Fixed